### PR TITLE
Save/reload last connected bot

### PR DESCRIPTION
**Pull request by CXuesong**
This will solve #118 , I suppose.

An option has been added to the settings dialog to enable this feature.

### CODE REVIEW

**Comment by kbremner** [comment]
Is this necessary?
Based on the discussion in #118, it seems like the solution to the actual problem is to persist settings between runs of the emulator?

**Comment by eanders-MS** [comment]
@CXuesong I appreciate your effort here but as @kbremner referenced, we're going to restore the bot connection on startup rather than prompt on exit.

**Comment by CXuesong** [comment]
Sorry, but perhaps I opened this PR in a hurry… So the last chosen bot will be selected on the startup now.

Still, uhm… may I keep the "prompt before exit" in the settings panel? I will remove it if it's not appropriate.

**Comment by eanders-MS** [comment]
Sorry for the wait. Unfortunately not going to be able to keep the "prompt before exit". Thanks for your understanding.

**Comment by CXuesong** [comment]
Okay, now we only have "restore last connected bot" feature.

Btw, I just found out that, if you delete a bot from the auto-completion list no sooner than you connected the bot (successfully or not), some `"DirectLine: Cannot get activities. No active bot"` will be shown repetitively in the console. Should [conversationsControllerV3.ts#L260](https://github.com/Microsoft/BotFramework-Emulator/blob/master/src/server/controllers/directLine/conversationsControllerV3.ts#L260) be commented out or changed to `log.trace`?

File: src/client/dialogs/appSettingsDialog.tsx [hunk 1]
```diff
@@ -71,7 +72,7 @@ export class AppSettingsDialog extends React.Component<{}, { ngrokPath: string }
     onAccept = () => {
         ServerSettingsActions.remote_setFrameworkServerSettings({
             ngrokPath: this.ngrokPathInputRef.value,
-            bypassNgrokLocalhost: this.bypassNgrokLocalhostInputRef.checked
+            bypassNgrokLocalhost: this.bypassNgrokLocalhostInputRef.checked,
```
Comment (eanders-MS) [inline src/client/dialogs/appSettingsDialog.tsx]: No need to add a comma here.

File: src/client/dialogs/appSettingsDialog.tsx [hunk 2]
```diff
@@ -84,16 +85,17 @@ export class AppSettingsDialog extends React.Component<{}, { ngrokPath: string }
...
-            defaultPath: dir,
-            properties: ['openFile']
-        }, (filenames: string[]) => {
-            if (filenames && filenames.length) {
-                // TODO: validate selection
-                this.ngrokPathInputRef.value = filenames[0];
-                this.setState({ ngrokPath: filenames[0] });
-            }
-        })
+                title: 'Browse for ngrok',
```
Comment (eanders-MS) [inline src/client/dialogs/appSettingsDialog.tsx]: Please undo reformatting of whitespace

File: src/client/dialogs/appSettingsDialog.tsx [hunk 3]
```diff
@@ -127,15 +129,15 @@ export class AppSettingsDialog extends React.Component<{}, { ngrokPath: string }
                             <li>
                                 <a href="javascript:void(0)"
                                     className={"emu-navitem emu-navitem-selected"}
-                                    >
+                                >
```
Comment (eanders-MS) [inline src/client/dialogs/appSettingsDialog.tsx]: There is a lot of whitespace-only modifications in this change. Please undo them.

**Comment by eanders-MS** [comment]
I'm sorry but I cannot accept this PR as-is. Something in it is buggy. Few incorrect behaviors I noticed:
- When I select a bot endpoint and click connect, my bot credentials disappear (expected), but then reappear (unexpected).
- Pressing the refresh button causes my bot's credentials to appear.

**Comment by CXuesong** [comment]
Oops… So now I have moved the bot reload logic to `componentDidMount`, as a callback for `serverSettings` event. Hope that solves the problem.

**Comment by eanders-MS** [comment]
Hi @CXuesong, Apologies for letting this languish, and let me explain the reason. We'd like to make a new emulator release before pulling in this feature so that it has adequate test time before going live. Additionally, we have a potentially conflicting change incoming, and would like to get that change in first, then merge this one.

**Comment by CXuesong** [comment]
Well, this is now a somewhat minor change, so I will try to follow up the branch and resolve the conflicts if possible ;-)

**Comment by eanders-MS** [comment]
Going ahead and closing this PR. Thank you for your effort, sorry it wasn't accepted.

**Pull request closed without merging**